components/org.wso2.carbon.identity.application.authenticator.adapter/src/main/java/org/wso2/carbon/identity/application/authenticator/adapter/AuthenticationRequestBuilder.java

```diff
@@ -0,0 +1,158 @@
+ * external authentication service.
+ */
+public class AuthenticationRequestBuilder implements ActionExecutionRequestBuilder {
+
+    private static final Log LOG = LogFactory.getLog(AuthenticationRequestBuilder.class);
+
+    @Override
+    public ActionType getSupportedActionType() {
+
+        return ActionType.AUTHENTICATION;
+    }
+
+    @Override
+    public ActionExecutionRequest buildActionExecutionRequest(Map<String, Object> eventContext)
+            throws ActionExecutionRequestBuilderException {
+
+        HttpServletRequest request = (HttpServletRequest) eventContext.get(AuthenticatorAdapterConstants.AUTH_REQUEST);
+        AuthenticationContext context = (AuthenticationContext) eventContext.get(
+                AuthenticatorAdapterConstants.AUTH_CONTEXT);
+
+        ActionExecutionRequest.Builder actionRequestBuilder = new ActionExecutionRequest.Builder();
+        actionRequestBuilder.flowId((String) eventContext.get(AuthenticatorAdapterConstants.FLOW_ID));
+        actionRequestBuilder.actionType(getSupportedActionType());
+        actionRequestBuilder.event(getEvent(request, context));
+        AllowedOperationBuilder allowedOperationBuilder = new AllowedOperationBuilder();
+        actionRequestBuilder.allowedOperations(allowedOperationBuilder.getAllowedOperations());
+
+        return actionRequestBuilder.build();
+    }
+
+    private Event getEvent(HttpServletRequest request, AuthenticationContext context)
+            throws ActionExecutionRequestBuilderException {
+
+        AuthenticatedUser lastAuthenticatedUser = context.getLastAuthenticatedUser();
+        String tenantDomain = context.getTenantDomain();
+
+        AuthenticationRequestEvent.Builder eventBuilder = new AuthenticationRequestEvent.Builder();
+        eventBuilder.tenant(new Tenant(String.valueOf(IdentityTenantUtil.getTenantId(tenantDomain)), tenantDomain));
+        if (lastAuthenticatedUser != null) {
+            eventBuilder.user(getUserForEventBuilder(lastAuthenticatedUser));
+            eventBuilder.organization(getOrganizationForEventBuilder(lastAuthenticatedUser));
+            eventBuilder.userStore(new UserStore(lastAuthenticatedUser.getUserStoreDomain()));
+        }
+        eventBuilder.currentStepIndex(context.getCurrentStep());
+        eventBuilder.authenticatedSteps(getAuthenticatedStepsForEventBuilder(context));
+        eventBuilder.request(getRequest(request));
+        return eventBuilder.build();
+    }
+
+    private User getUserForEventBuilder(AuthenticatedUser authenticatedUser)
+            throws ActionExecutionRequestBuilderException {
+
+        try {
+            return new AuthenticationRequestUser(authenticatedUser.getUserId(), authenticatedUser);
+        } catch (UserIdNotFoundException e) {
+            throw new ActionExecutionRequestBuilderException("User ID not found for current authenticated user.", e);
+        }
```

Comment:
If user id is not present let's proceed with user name rather than throwing the error.

Reply:
Had offline discussion with Malithi and decide that user id must be mandatory to identify the local user.